controller/volume_controller.go

```diff
@@ -1026,6 +1031,116 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, es map[stri
 	return nil
 }
 
+func (vc *VolumeController) ReconcileLocalReplica(v *longhorn.Volume, e *longhorn.Engine, rs map[string]*longhorn.Replica) error {
+	if v.Spec.DataLocality != types.DataLocalityBestEffort {
+		return nil
+	}
+
+	if v.Status.State != types.VolumeStateAttached {
+		return nil
+	}
+
+	if e == nil {
+		return fmt.Errorf("BUG: ReconcileLocalReplica needs a valid engine")
+	}
+
+	if !vc.isSafeToModifyReplica(v, e, rs) {
+		return nil
+	}
+
+	if !hasLocalReplicaOnSameNodeAsEngine(e, rs) {
+		if _, err := vc.addLocalReplicaOnSameNodeAsEngine(v, e, rs); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// check if there are too many healthy replicas
+	healthyCount := 0
+	for _, r := range rs {
+		if r.Spec.FailedAt == "" && r.Spec.HealthyAt != "" {
+			healthyCount++
+		}
+	}
+
+	if healthyCount <= v.Spec.NumberOfReplicas {
+		return nil
+	}
+
+	// Randomly delete extra non-local healthy replicas
```

Comment:
Can we make this cleanup more intelligent? e.g., if soft anti-affinity is enabled and there are multiple replicas on the same node, deleting one of those replicas would be better

Reply:
Thank you! I will try this

Reply:
@yasker How do you think about this?

Reply:
Yes, we should do this. Though I cannot think of a case other than soft anti-affinity we need to cover.

Reply:
So, we don't care about zone and we just consider replicas on the same node/disk as better candidates for deletion?

Reply:
@shuo-wu @yasker I updated replica cleanup logic. How do you think about this logic?

Reply:
LGTM.